### PR TITLE
Pass zone to backends

### PR DESCRIPTION
**Pull request by AwesomePatrol**
As described in https://github.com/cilium/design-cfps/pull/31, this is part of https://github.com/cilium/cilium/issues/31752

`Zone` string is taken from `EndpointSlice`, then copied to `Backend` which is later reflected in BPF map as uint8 id in `lb{4,6}_backend`. uint8<->string mapping is based on configuration passed in `fixed-zone-mapping`. It follows the same pattern as `fixed-identity-mapping`.

Example configuration (in `cilium-config`) for KIND cluster:
```
  fixed-zone-mapping: zone-a=123,zone-b=129
```
where topology labels were set on the nodes:
```
kubectl label node kind-worker topology.kubernetes.io/zone=zone-a
kubectl label node kind-worker topology.kubernetes.io/zone=zone-b
```

Can be verified with `cilium map get cilium_lb4_backends_v3`:
```
Key   Value                        State   Error
17    ANY://10.244.0.151[zone-b]   sync
13    ANY://192.168.11.2[zone-a]   sync
18    ANY://10.244.1.64[zone-a]    sync
19    ANY://10.244.1.64[zone-a]    sync
14    ANY://10.244.1.70[zone-a]    sync
15    ANY://10.244.1.70[zone-a]    sync
16    ANY://10.244.0.151[zone-b]   sync
```

```release-note
Store zone information alongside lb{4,6}_backend entries (based on mapping from fixed-zone-mapping and values from EndpointSlices)
```

### CODE REVIEW

File: pkg/zone/zone.go
Comment (joamaki) [inline pkg/zone/zone.go]: s/siltenly/silently/

File: pkg/zone/zone.go
Comment (joamaki) [inline pkg/zone/zone.go]: Don't leave TODOs in the code, but rather create a github issue that details why this is needed to be done in the future.

**Review by joamaki — APPROVED** [review]

File: pkg/zone/zone.go
Comment (AwesomePatrol) [reply pkg/zone/zone.go]: This implementation was replaced with `FixedZoneMapping` added in the later commit. Would you prefer me to squash them?

File: pkg/k8s/endpoints.go
Comment (tklauser) [inline pkg/k8s/endpoints.go]: nit: could use `k8s.io/api/core/v1.LabelTopologyZone` here instead of the verbatim string `topology.kubernetes.io/zone`

File: pkg/k8s/endpoints.go
Comment (tklauser) [inline pkg/k8s/endpoints.go]: same here, could use `k8s.io/api/core/v1.LabelTopologyZone`

**Review by tklauser — APPROVED** [review]
Looks good for my codeowners, small nits only. Thanks.

File: pkg/k8s/endpoints.go
Comment (AwesomePatrol) [reply pkg/k8s/endpoints.go]: In this file `github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1` is used. Is it ok to import both or should I replace slim version with the regular one? Or copy these constants to the slim version?

Comment (tklauser) [reply pkg/k8s/endpoints.go]: I think it's fine to import both, we already import both in the `github.com/cilium/cilium/pkg/k8s` package.

Comment (AwesomePatrol) [reply pkg/k8s/endpoints.go]: Thanks for quick reply. Done.

I also replaced hostname label reference with const for consistency in this file

**Review by aspsk — APPROVED** [review]
Datapath part looks good

**Review by learnitall — APPROVED** [review]
API changes look good :rocket:, thanks!

**Comment by AwesomePatrol** [comment]
@ysksuzuki, please leave a comment or reassign as your review is required to unblock this PR. Thanks!

File: pkg/zone/zone.go
Comment (christarazi) [inline pkg/zone/zone.go:1]: I think we might as well squash this commit with the one before it, if we are going to create a new package just to replace it in the next commit.

**Review by ysksuzuki — APPROVED** [review]
LGTM. Thanks!

File: pkg/zone/zone.go
Comment (AwesomePatrol) [reply pkg/zone/zone.go:1]: Done

**Pull request merged**
